File: dart/day7_part1_2018.dart

import 'dart:io';

void main() {
  var lines = File('input.txt').readAsLinesSync();
  var deps = <String, List<String>>{};
  var allSteps = <String>{};

  for (var line in lines) {
    var a = line[5];
    var b = line[36];
    deps.putIfAbsent(b, () => []).add(a);
    allSteps.addAll([a, b]);
  }

  var order = topologicalSort(deps, allSteps);
  print(order);
}

String topologicalSort(Map<String, List<String>> deps, Set<String> allSteps) {
  var order = <String>[];
  var available = <String>[];

  for (var step in allSteps) {
    if (deps[step] == null || deps[step]!.isEmpty) {
      available.add(step);
    }
  }
  available.sort();

  while (available.isNotEmpty) {
    var next = available.removeAt(0);
    order.add(next);

    for (var step in allSteps) {
      if (deps[step]?.contains(next) ?? false) {
        deps[step]!.remove(next);
        if (deps[step]!.isEmpty) {
          available.add(step);
        }
      }
    }
    available.sort();
  }
  return order.join();
}

File: kotlin/day7_part1_2018.kt
import java.io.File

fun main(args: Array<String>) {
    val (deps, allSteps) = parseInput("input.txt")
    val order = topologicalSort(deps, allSteps)
    println(order)
}

fun parseInput(filename: String): Pair<MutableMap<Char, MutableList<Char>>, MutableMap<Char, Boolean>> {
    val deps = mutableMapOf<Char, MutableList<Char>>()
    val allSteps = mutableMapOf<Char, Boolean>()

    File(filename).forEachLine {
        val parts = it.split(" ")
        val a = parts[1].single()
        val b = parts[7].single()
        deps.computeIfAbsent(b) { mutableListOf() }.add(a)
        allSteps[a] = true
        allSteps[b] = true
    }

    return Pair(deps, allSteps)
}

fun topologicalSort(deps: MutableMap<Char, MutableList<Char>>, allSteps: MutableMap<Char, Boolean>): String {
    val order = mutableListOf<Char>()
    val available = mutableListOf<Char>()

    for (step in allSteps.keys) {
        if (deps[step].isNullOrEmpty()) {
            available.add(step)
        }
    }
    available.sort()

    while (available.isNotEmpty()) {
        val next = available.removeAt(0)
        order.add(next)

        for (step in allSteps.keys) {
            if (deps[step]?.contains(next) == true) {
                deps[step]?.remove(next)
                if (deps[step].isNullOrEmpty()) {
                    available.add(step)
                }
            }
        }
        available.sort()
    }
    return order.joinToString("")
}